python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -1175,21 +1162,109 @@ class PyInferenceRequest
     ThrowIfError(TRITONSERVER_InferenceRequestSetDoubleParameter(
         triton_object_, key.c_str(), value));
   }
+
   void Cancel()
   {
     ThrowIfError(TRITONSERVER_InferenceRequestCancel(triton_object_));
   }
 
+  // Response management
+  void GetNextResponse(const py::object& py_future)
+  {
+    // Called exclusively from Python, so GIL is always held.
+    std::lock_guard lock(response_mu_);
 
- public:
-  std::unique_ptr<CallbackResource> request_callback_resource_{nullptr};
+    // There are two cases:
+    // 1. The backend is faster than the frontend:
+    //      There will be response(s) queued at 'responses_' deque, so pop the
+    //      response from the deque and set it to the 'py_future' object.
+    // 2. The backend is slower than the frontend:
+    //      The 'responses_' deque will be empty, so track the 'py_future' in
+    //      this request object until the 'AddNextResponse()' is called.
+    if (responses_.empty()) {
+      if (response_future_.ptr() != nullptr) {
+        throw AlreadyExistsError("cannot call GetNextResponse concurrently");
+      }
+      response_future_ = py_future;
+    } else {
+      std::pair<std::shared_ptr<PyInferenceResponse>, const uint32_t>&
+          py_response = responses_.front();
+      PyFutureSetResult(py_future, py_response);
+      responses_.pop_front();
+    }
+  }
 
  private:
-  std::unique_ptr<PyResponseAllocator::CallbackResource>
-      allocator_callback_resource_{nullptr};
-  std::unique_ptr<PyInferenceResponse::CallbackResource>
-      response_callback_resource_{nullptr};
+  void AddNextResponse(
+      struct TRITONSERVER_InferenceResponse* response, const uint32_t flags)
+  {
+    // Pair the 'response' and 'flags' so they can be easily moved around.
+    std::shared_ptr<PyInferenceResponse> managed_ptr(nullptr);
+    if (response != nullptr) {
+      managed_ptr.reset(new PyInferenceResponse(response, true /* owned */));
+    }
+    std::pair<std::shared_ptr<PyInferenceResponse>, const uint32_t> py_response(
+        std::move(managed_ptr), std::move(flags));
+
+    // There are two cases:
```

Comment:
```suggestion
    // There are two cases:
    // 1. GetNextResponse() hasn't been called and the future is not set.
    // 2. GetNextResponse() has been called and future is set.
```

Reply:
yes, the comments are updated. [Update comments related to GetNextResponse() and AddNextResponse()](https://github.com/triton-inference-server/core/pull/426/commits/1d4f98df4530d1e0b75ee426f33ff80d035840c0)